### ISSUE TITLE
haxe.Serializer and arrays

### ISSUE THREAD

**Issue by aartzrc** [HaxeFoundation/haxe#8314]
I ran into an issue with arrays that appears to be C# only. When using `haxe.Unserializer` the array appears to become read only. Here's an example:

``` haxe
var simpleObj:{val:Array<Float>} = { val: [ 0, 0, 0, 0 ] };
var newVal1:Int = 1;
var newVal2:Int = 2;

simpleObj.val[0] = newVal1;
trace(simpleObj);

var afterSerialize:{val:Array<Float>} = Unserializer.run(Serializer.run(simpleObj));
afterSerialize.val[0] = newVal2;
trace(afterSerialize);

afterSerialize.val = afterSerialize.val.copy();
afterSerialize.val[0] = newVal2;
trace(afterSerialize);
```

In interp, js, etc I get what is expected:
```
{val: [1,0,0,0]}
Class<Array>
{val: [2,0,0,0]}
Class<Array>
{val: [2,0,0,0]}
Class<Array>
```
In C# I get the following - notice it doesn't update the value until I have made a copy of the array and re-assigned it:
```
{ val : [1,0,0,0] }
Array`1[System.Double]
{ val : [1,0,0,0] }
Array`1[System.Double]
{ val : [2,0,0,0] }
Array`1[System.Double]
```
No run-time errors, it just ignores the assignment. Haxe RC2, Net framework V4.6.2 and Visual Studio 2019.

**Comment by RealyUniqueName** [comment]
Nice. Thanks!

**Comment by ncannasse** [comment]
The problem must be that everytime the Array is read as an Array<Int> it's casted from Array<Dynamic> to Array<Int> and this creates a new Array whereas this should only be done once and then should "fix" the Array type.

This should be reproducible by creating a new Array<Dynamic> and put it in a Dynamic variable then assign it to an Array<Int>. I think we need more unit tests to take into account these cases.

**Comment by aartzrc** [comment]
This appears to have gotten 'worse' with RC3, although it is more apparent what is happening so easier to track down.

The point that things break down appears to be when working with anonymous objects during the `__hx_cast` call. In RC3 the fields become `Array<string>` vs `Array<object>`, the test to determine if the current array field should be returned happens in `Array.Array_cast`, the call to `haxe.lang.Runtime.eq` returns false and a new array is built causing things to become 'detached'.

We have a transaction system that creates a copy of the original data on demand, I haven't had time to build a full test around it yet but here's some snippets of output:

```haxe
var newParent = transaction.get(parentGuid); // This creates a fresh anonymous object
var t = newParent.children; // Keep track of the 'children' array - by reference
newParent.children.push(childGuid); // Add an item to 'children'
// t should contain 'childGuid' now, RC3 in C# this fails
```

```cs
// RC2 CS output - this works
object newParent = global::data.store.RecordDataHelpers.get(transaction, parentGuid);
var t = ((global::Array<object>)(global::Array<object>.__hx_cast<object>(((global::Array)(global::haxe.lang.Runtime.getField(newParent, "children", 1649795810, true))))));
((global::Array<object>) (global::Array<object>.__hx_cast<object>(((global::Array) (global::haxe.lang.Runtime.getField(newParent, "children", 1649795810, true)) ))) ).push(newVal);
```

```cs
// RC3 CS output - this fails, probably due to conversion to Array<string> breaking object references?
object newParent = global::data.store.RecordDataHelpers.get(transaction, parentGuid);
var t = ((global::Array<string>) (global::Array<object>.__hx_cast<string>(((global::Array) (global::haxe.lang.Runtime.getField(newParent, "children", 1649795810, true)) ))) );
((global::Array<string>) (global::Array<object>.__hx_cast<string>(((global::Array) (global::haxe.lang.Runtime.getField(newParent, "children", 1649795810, true)) ))) ).push(newVal);
```

**Comment by kLabz** [comment]
#8387 is definitely responsible for these changes.
If you can give us some code to reproduce the issue, I might be able to help. 

There is also an [open?] issue regarding array references lost during cast, but I cannot find it atm.

**Comment by aartzrc** [comment]
Thanks @kLabz, very happy to know someone is able to work on this one.

I narrowed it down to the same concept as the original issue. We are serializing/deserializing for data storage and after some tests it appears to be directly related to the serialization process. I think the issue was around in RC2 but has progressed into other parts of the C# output now.

The code below is essentially the same as first submitted but uses `haxe.Json` serialization instead. It works in RC2 but fails in RC3.

The original issue was for `haxe.Serializer` which began failing between RC1 and RC2. That may help track things down, they look related to me.

Regular fields are ok, but array fields cause trouble. I haven't been able to track down the specific thing that causes anonymous object arrays to be different before and after serialization.

```haxe
class Main  {
    static function main() {
        // Test 1
        var t:DataItem = { data:"d", a:[] };

        // Add to the anonymous object array - this works
        var a = t.a;
        t.a.push("raw anonymous object is ok");
        trace(a);
        trace(t.a);


        // Test 2
        var t:DataItem = haxe.Json.parse(haxe.Json.stringify({ data:"d", a:[] }));
        
        // Add to the anonymous object array - fails after parse
        var a = t.a;
        t.a.push("after json conversion fails");
        trace(a);
        trace(t.a);
        t.data = "simple fields are ok, just arrays fail?";
        trace(t);
    }
}

typedef DataItem = {data:String, a:Array<String>};
```

**Comment by kLabz** [comment]
JsonParser generates `Array<Dynamic>` for every array, and I suppose Unserializer works in a similar way. As Nicolas said, casting the array make it loose reference here..

The only thing my PR changed was that this now applies to `Array<String>` **too** (since before, `String` was used as `object` it did not have to to any cast from `Array<Dynamic>`).

Does this make this issue a "duplicate" of #4872 @Simn ?

**Comment by Simn** [comment]
Yes it probably is, but this still means that we're dealing with an unfortunate regression now...

**Comment by kLabz** [comment]
Well there's already the issue with other basic types (the above example fails too (and failed with 3.4.x, rc1 and rc2) with `Array<Int>` and probably with anything other than `Array<Dynamic>`)

**Comment by nadako** [comment]
I don't know, #2103 was there for ages and this is just an instance of that. Yes, technically it's a regression for `String` parameters, but if we use some other basic type, the problem will be there on earlier versions too.

In my previous project I worked this around by using `-D erase-generics`, although since then I realized that we're actually trying to assign `Array<Dynamic>` to `Array<OtherType>` here, and this is not safe (Haxe won't normally allow it without a `cast`/`Dynamic`), so one could expect issues. The correct way would be to post-process the unserialized data, or change the serializer to create properly typed array right away (e.g. by providing some callback to it).

UPD: That said, a `cast` still should NOT create new objects :)

**Comment by aartzrc** [comment]
After reading through previous issues related to this it looks like covering all scenarios on static targets is very difficult. The balance between maintaining runtime efficiency (cast to a specific type) and code flexibility+reference integrity (never cast and keep all arrays dynamic) is tricky.

Internally we use `typedef`s for everything and only use reflection to handle data migration from a previous `typedef` so we generally know the specific array type ahead of time. The compiler should be able to see this and properly cast during parsing. As a side note `tink.Json` creates the typed arrays during parse and does not have the 'cast on the fly' issue we're seeing here, but it creates quite a bit of extra code and will drop fields not explicitly defined (there are work-arounds available of course!)

For us, a way to target specific `typedef` fields and have them auto-convert to `Array<Any>` after other type checking is done would give us a quick solution. It would allow us to maintain type safety in our code, but then the compiler would output `Array<object>` as previously and avoid the cast copy problem. Some runtime slow down, but we'd be aware of it - macro anybody?

Just for reference, here's a few more ideas/tests done to see how the dynamic objects react:
```haxe
class Main  {
    static function main() {
        var t:DataItem = haxe.Json.parse(haxe.Json.stringify({ data:"d", strArray:[], intArray:[], dynArray:[], anyArray:[] }));
        
        trace("getField cast looses reference");
        var strArray = t.strArray;
        strArray.push("via reference");
        t.strArray.push("via getField");
        trace(strArray);
        trace(t.strArray);

        trace("getField WITH setField 'fixes' reference for future getField calls");
        var strArray = t.strArray;
        t.strArray = strArray; // Re-assign fixes (setField call)
        strArray.push("via reference");
        t.strArray.push("via getField");
        trace(strArray);
        trace(t.strArray);

        var intArray = t.intArray;
        t.intArray = intArray; // Re-assign fixes (setField call)
        intArray.push(0);
        t.intArray.push(1);
        trace(intArray);
        trace(t.intArray);

        trace("Array<Dynamic> does not have this problem");
        var dynArray = t.dynArray;
        dynArray.push(0);
        dynArray.push("test");
        trace(dynArray);
        trace(t.dynArray);

        trace("Array<Any> does not have this problem");
        var anyArray = t.anyArray;
        anyArray.push(0);
        anyArray.push("test");
        trace(anyArray);
        trace(t.anyArray);

        trace("Converting to Array<Any> requires a cast by compiler, and it is more apparent that a copy is being made and reference is lost");
        var anyArray:Array<Any> = cast intArray;
        anyArray.push("test");
        trace(anyArray);
        trace(intArray);

        trace("Converting to Array<Dynamic> fails at run time - may be OK?");
        var dynArray:Array<Dynamic> = intArray;
        dynArray.push("test");
        trace(dynArray);
        trace(intArray);
    }
}

typedef DataItem = {data:String, strArray:Array<String>, intArray:Array<Int>, dynArray:Array<Dynamic>, anyArray:Array<Any>};

```

**Comment by nadako** [comment]
Yeah, that's exactly what I'm talking about - you either have a type-aware parser (like `tink_json`) that knows what type to create, or you post-process the dynamic data. Your quick solution is good enough for your case I guess and you can macro-generate the code that does this reassigning, that's a nice macro exercise :)

Another idea for experimenting is hack into `haxe.format.JsonParser` and make it receive some kind of schema structure in additional to the String, so it knows what exact types to construct, and then macro-generate your schema from type definitions. Compared to `tink_json`, this would be a kind of middle-ground between code-size and run-time performance, I guess.

**Comment by hoseyjoe** [comment]
Isn't this an issue for all static targets including Hxcpp and Hashlink (It is not labeled for those)?  Maybe those 2 have a workaround already? It would be better to have fix in the core vs each static target having its own fix.

**Comment by kLabz** [comment]
HL should have been fixed last month, I don't know about cpp. The general issue is tracked in #4872

**Comment by hoseyjoe** [comment]
There hasn't been any motion on #4872.  Simon said, in Dec 2018, that it was moved to 4.1 release. So I am assuming it might get looked at after 4 is official?  So best case 6 months to a year?  I am just trying to evaluate how much patching I have to do in the short run and I appreciate everyone's work/information

**Comment by aartzrc** [comment]
Just wondering, could the output be changed to use the C# `dynamic` type? Instead of the `Runtime.getField`/`Runtime.setField` calls with `__hx_cast` mixed in they would use the C# object directly, the output would look similar to Javascript and the Haxe compiler in theory would never make a bad cast that would be a runtime bug.

Something like this:
```cs
dynamic newParent = global::data.store.RecordDataHelpers.get(transaction, parentGuid);
var t = (global::Array<dynamic>)newParent.children;
((global::Array<dynamic>)newParent.children).push(newVal);
```

**Comment by RealyUniqueName** [comment]
Since it's a regression I'm gonna take a look at this for 4.0.

**Comment by RealyUniqueName** [comment]
Ok, doesn't look like this could be fixed without reverting #8387 which would bring #8361 back.
And using `dynamic` wouldn't help because on passign that array to any `Array<String>` place C# runtime will emit an exception.

So, it's either we close this in favor of #4872 or revert #8387. I'm leaning towards the former.

**Comment by RealyUniqueName** [comment]
Closing in favor of #4872.
